### PR TITLE
Add sapcar extraction fns

### PR DESCRIPTION
**Pull request by Simranpal**
Adding the sapcar extraction functionality for SAP car/sar files.
Current function is for extraction only, as we don't need the other functionalities(like viewing, compressing archives). SAPCAR utility has lot of extra options, which is why I have added `options` to support additional parameters.

### CODE REVIEW

**Comment by Simranpal** [comment]
image

File: shaptools/shell.py [hunk 1]
```diff
@@ -187,3 +191,25 @@ def remove_user(user, force=False, root_user=None, root_password=None, remote_ho
...
+    Args:
+        sar_file (str): Path to the sar file to be extracted
+        options(str, opt): Additional options to SAPCAR command
+        user (str, opt): User to execute the SAPCAR command
+        password (str, opt): User password
+    """
+    if not os.path.isfile(sar_file):
+        raise FileDoesNotExistError('The SAR file \'{}\' does not exist'.format(sar_file))
+    if options is None:
+        cmd = './SAPCAR -xvf {sar_file}'.format(sar_file=sar_file)
```
Comment (diegoakechi) [inline shaptools/shell.py]: I would have the option `-manifest SIGNATURE.SMF` enabled by default as it generates the manifest used by the installer. Without it the installation fails.

File: shaptools/shell.py [hunk 2]
```diff
@@ -187,3 +191,25 @@ def remove_user(user, force=False, root_user=None, root_password=None, remote_ho
         else:
             break
     raise ShellError('error removing user {}'.format(user))
+
+def extract_sapcar_cmd(sar_file, options=None, user=None, password=None, remote_host=None):
```
Comment (diegoakechi) [inline shaptools/shell.py]: I would call it `extract_sapcar_file`

File: shaptools/netweaver.py [hunk 3]
```diff
@@ -416,3 +416,20 @@ def stop(self, wait=15, delay=0, exception=True, **kwargs):
         if exception and result.returncode:
             raise NetweaverError('Error running sapcontrol command: {}'.format(result.cmd))
         return result
+
+    def extract_sap_installer(
```
Comment (diegoakechi) [inline shaptools/netweaver.py]: I don't think we need special function for each SAR file. The extraction is a genetic process. For me it is fine to raise an generic extraction error pointing to the file.

File: shaptools/shell.py [hunk 4]
```diff
@@ -23,6 +23,10 @@ class ShellError(Exception):
     Exceptions in shell module
     """
 
+class FileDoesNotExistError(Exception):
```
Comment (arbulu89) [inline shaptools/shell.py]: Add double new lines between class definitions. By definition:
- 2 new lines between methods outside classes
- 1 new line in method inside classes

File: shaptools/shell.py (hunk 1 above)
Comment (arbulu89) [inline shaptools/shell.py]: Having `./SAPCAR` looks a bit risky, right?
Why do you guess that the file is in the current directory? Actually, as this code is executed in python `.` folder might be considered the python execution file location, what it doesn't look where `SAPCAR` is located.
We need to provide an extra variable to say where the `SAPCAR` executable is located.

File: shaptools/shell.py (hunk 2 above)
Comment (arbulu89) [inline shaptools/shell.py]: I don't really like to have so many parameters with `None` default value.
Use `kwargs` to bring them. It looks like we might need to include more things in the future, so the optional parameters are better in `kwargs` here.

File: shaptools/shell.py (hunk 1 above)
Comment (arbulu89) [reply shaptools/shell.py]: Is this always the case? I wouldn't like to have hardcoded things here. What if you don't want to use this option?
I would let this as it is, and set this option in higher levels (salt code for example)

File: shaptools/hana.py [hunk 5]
```diff
@@ -701,3 +701,19 @@ def unset_ini_parameter(
             file_name=file_name, layer=layer, layer_name=layer_name,
             set_value=False, reconfig=reconfig, key_name=key_name,
             user_name=user_name, user_password=user_password)
+
+    def extract_hdbserver(self, sar_file, options=None, user=None, password=None, remote_host=None):
```
Comment (arbulu89) [inline shaptools/hana.py]: You don't need to pass `user`, `password` and `remote_host` options. They are already part of the class. You can retrieve them using `self` option.

File: shaptools/netweaver.py [hunk 6]
```diff
@@ -416,3 +416,20 @@ def stop(self, wait=15, delay=0, exception=True, **kwargs):
...
+        If user and password are provided it will be executed with this user.
+
+        Args:
+            sar_file (str): Path to the sar file to be extracted
+            options(str, opt): Additional options to SAPCAR command
+            user (str, opt): User to execute the SAPCAR command
+            password (str, opt): User password
+        """
+        result = shell.extract_sapcar_cmd(sar_file, options, user, password, remote_host)
+        if result.returncode:
```
Comment (arbulu89) [inline shaptools/netweaver.py]: This is not possible, right? I mean, `shell.extract_sapcar_cmd` will already raise an exception if `returncode` is different than `0`.
Here, if you want to re raise a new exception, you will need to use the `try/catch` option.

File: shaptools/hana.py [hunk 7]
```diff
@@ -701,3 +701,19 @@ def unset_ini_parameter(
...
+        If user and password are provided it will be executed with this user.
+
+        Args:
+            sar_file (str): Path to the sar file to be extracted
+            options(str, opt): Additional options to SAPCAR command
+            user (str, opt): User to execute the SAPCAR command
+            password (str, opt): User password
+        """
+        result = shell.extract_sapcar_cmd(sar_file, options, user, password, remote_host)
+        if result.returncode:
```
Comment (arbulu89) [inline shaptools/hana.py]: This is not possible, right? I mean, shell.extract_sapcar_cmd will already raise an exception if returncode is different than 0.
Here, if you want to re raise a new exception, you will need to use the try/catch option.

File: shaptools/hana.py (hunk 5 above)
Comment (arbulu89) [reply shaptools/hana.py]: And I wonder, if we need these methods. They don't do anything, just call again other method without adding any special thing...
If we can, I would remove them, they add not needed complexity

File: shaptools/netweaver.py (hunk 6 above)
Comment (Simranpal) [reply shaptools/netweaver.py]: Yes thats correct, I will raise the generic exception

File: shaptools/shell.py (hunk 2 above)
Comment (Simranpal) [reply shaptools/shell.py]: Yes it makes sense to use `kwargs` here

File: shaptools/shell.py (hunk 1 above)
Comment (Simranpal) [reply shaptools/shell.py]: It was explained in the requirement, that SAPCAR should be present in current directory.
But we need to make it flexible as you explained, I will pass the extra variable for the SAPCAR executable path

File: shaptools/saputils.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+:author: sisingh
+:organization: SUSE LLC
+:contact: sisingh@suse.com
+
+:since: 2020-03-12
+"""
+#TODO: Add support for other SAPCAR functionalties apart from extraction
+
+from __future__ import print_function
```
Comment (arbulu89) [inline shaptools/saputils.py]: I don't think we need this import, right?

File: shaptools/saputils.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+def extract_sapcar_file(sapcar_exe, sar_file, **kwargs):
+    """
+    Execute SAPCAR command to decompress a SAP CAR or SAR archive files.
+    If user and password are provided it will be executed with this user.
+
+    Args:
+        sapcar_exe(str): Path to the SAPCAR executable
+        sar_file (str): Path to the sar file to be extracted
+        options (str, opt): Additional options to SAPCAR command
+        extract_dir (str, opt): Directory where archive will be extracted. It creates the dir
```
Comment (arbulu89) [inline shaptools/saputils.py]: I would call it `output_dir`, but this option is good as well

File: shaptools/saputils.py [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    if not os.path.isfile(sar_file):
+        raise FileDoesNotExistError('The SAR file \'{}\' does not exist'.format(sar_file))
+
+    options = kwargs.get('options', None)
+    extract_dir = kwargs.get('extract_dir', None)
+    user = kwargs.get('user', None)
+    password = kwargs.get('password', None)
+    remote_host = kwargs.get('remote_host', None)
+
+    extract_dir_str = '-R {}'.format(extract_dir) if extract_dir else ''
```
Comment (arbulu89) [inline shaptools/saputils.py]: I would change `-R {}` by ` -R {}` , notice the first space character.

File: shaptools/saputils.py [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+    options_str = '{}'.format(options) if options else ''
```
Comment (arbulu89) [inline shaptools/saputils.py]: The same, add an initial space character here

File: shaptools/saputils.py [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+
+    cmd = '{sapcar_exe} -xvf {sar_file} {options_str} {extract_dir_str}'.format(
```
Comment (arbulu89) [inline shaptools/saputils.py]: If you add the commented spaces in the previous lines, we could change this to:
`{sapcar_exe} -xvf {sar_file}{options_str}{extract_dir_str}'`
This way we wouldn't have strange space characters hanging there

File: shaptools/saputils.py (hunk 8 above)
Comment (Simranpal) [reply shaptools/saputils.py]: Yes, I used it for testing but can remove this

File: shaptools/saputils.py (hunk 12 above)
Comment (Simranpal) [reply shaptools/saputils.py]: Yes I guess I can do it this way

**Comment by diegoakechi** [comment]
@Simranpal @arbulu89 a suggestion here is to make the fn generic like execute_sapcar the same for the states, and let the formula specialize the call that it needs. This way we follow similar approach salt has on the archive states. What do you think?

File: shaptools/shell.py (hunk 1 above)
Comment (Simranpal) [reply shaptools/shell.py]: as discussed here, I have set this option from formula level as `options` parameter

File: tests/saputils_test.py [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+
+    @classmethod
+    def tearDownClass(cls):
+        """
+        Global tearDown.
+        """
+
+    @mock.patch('shaptools.shell.execute_cmd')
+    @mock.patch('os.path.isfile')
+    def test_extract_sapcar_file(self, mock_sapcar_file, mock_execute_cmd):
```
Comment (arbulu89) [inline tests/saputils_test.py]: Please, test the `output_dir` and `options` entries.

Comment (Simranpal) [reply tests/saputils_test.py]: Thanks, I added those options now

File: tests/saputils_test.py [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+
+# pylint:disable=C0103,C0111,W0212,W0611
+
+import os
+import sys
+sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))
+
+import logging
+import unittest
+import filecmp
```
Comment (arbulu89) [inline tests/saputils_test.py]: @Simranpal We don't need to import `filecmp` and `shutil`. Please remove them

Comment (Simranpal) [reply tests/saputils_test.py]: Thanks, done!

**Pull request merged**